dhall/src/Dhall/Eval.hs

```diff
@@ -427,6 +427,9 @@ vWith (VRecordLit kvs) (k₀ :| k₁ : ks) v = VRecordLit (Map.insert k₀ e₂
             Just e₁' -> e₁'
 
     e₂ = vWith e₁ (k₁ :| ks) v
+vWith (VNone _T) ("?" :| _      ) _ = VNone _T
+vWith (VSome  _) ("?" :| []     ) v = VSome v
+vWith (VSome  t) ("?" :| k₁ : ks) v = VSome (vWith t (k₁ :| ks) v)
```

Comment:
I think this feels a bit stringly-typed. A more robust solution would probably turn the

`(NonEmpty Text)` field from the `With` constructor into something like `NonEmpty WithKey` where

```
data WithKey = Label Text | QuestionMark
```

But maybe this is overkill. So far the logic involving the special `?` label is very localized.